### PR TITLE
On ScaleDown remove old/stale SSM parameters

### PR DESCRIPTION
**Pull request by jeanschmidt**

### CODE REVIEW

**Review by zxiiro — APPROVED** [review]
Seems reasonable to me although I'm probably not the best person to review this since I don't fully understand how the SSM parameters are used in the ALI.

**Comment by jeanschmidt** [comment]
Thanks for the trust @zxiiro, but lets be very mindful to not stamp javascript code without reflecting the changes in the tests. We strive for 100% coverage and give lots of trust to developers to include changes.

We can't do that with terraform, but we can be mindful with typescript :)

**Comment by jeanschmidt** [comment]
Screenshot 2024-07-23 at 20 18 00

**Review by ZainRizvi — COMMENTED** [review]
A few questions here:
1. Do we know what these are used for?
2. What triggers their creation?
3. We only have 54 SSM keys right now unless I looking in the wrong spot (and please let me know if I am!). That puts us within their free tier.  Even if that grows 100x in the next two months, the bill is still going to be in the 2-digit dollar amount.  This feels like a potentially risky change (at least without the above two questions answered).  Is this really worth experimenting with compared to the more immediate problems we have to solve?

Basically, I want to make sure we don't accidentally break LF runners with this change. But I also don't want to spend too much eng time on what appears to be a very low-value task

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts [hunk 1]
```diff
@@ -225,6 +239,46 @@ export async function scaleDown(): Promise<void> {
...
+        Array.from((await listSSMParameters(metrics, awsRegion)).values()),
+      );
+
+      let deleted = 0;
+      for (const ssmParam of ssmParams) {
+        /* istanbul ignore next */
+        if (ssmParam.Name === undefined) {
+          continue;
+        }
+        if (ssmParam.LastModifiedDate === undefined) {
```
Comment (ZainRizvi) [inline terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts:255]: same q as earlier about undefined dates.  Whatever you decide is the right thing to do here, it would be helpful to future maintainers to leave a comment explaining the reasoning

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/config.ts [hunk 2]
```diff
@@ -98,6 +100,8 @@ export class Config {
     this.scaleConfigRepo = process.env.SCALE_CONFIG_REPO || 'test-infra';
     this.scaleConfigRepoPath = process.env.SCALE_CONFIG_REPO_PATH || '.github/scale-config.yml';
     this.secretsManagerSecretsId = process.env.SECRETSMANAGER_SECRETS_ID;
+    this.sSMParamCleanupAgeDays = Number(process.env.SSM_PARAM_CLEANUP_AGE_DAYS || '7');
```
Comment (ZainRizvi) [inline terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/config.ts:103]: Don't we occasionally have runners that live for longer than 7 days?  Perhaps we should be more conservative about this time range to account for worst case possibilities

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts [hunk 3]
```diff
@@ -225,6 +239,46 @@ export async function scaleDown(): Promise<void> {
...
+        if (
+          ssmParam.LastModifiedDate.getTime() >
+          moment().subtract(Config.Instance.sSMParamCleanupAgeDays, 'days').toDate().getTime()
+        ) {
+          break;
+        }
+        if (await doDeleteSSMParameter(ssmParam.Name, metrics, awsRegion)) {
+          deleted += 1;
+        }
+        if (deleted >= Config.Instance.sSMParamMaxCleanupCount) {
```
Comment (ZainRizvi) [inline terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts:267]: Looks like `sSMParamMaxCleanupCount` is the max number of params that should get cleaned up on a single scale down request.  However, when ten instances are being scaled down one after another, you could end up deleting 10 * `sSMParamMaxCleanupCount` instances by the end.

Is that right?

If so, I'm wondering if we can change this name to reflect that expectation.  For context, my first assumption was that this var represented the threshold where we would only cleanup SSM params if we had more than `sSMParamMaxCleanupCount` entries

Potential new name: sSMParamCleanupAllowancePerScaledown

**Review by ZainRizvi — APPROVED** [review]

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts [hunk 4]
```diff
@@ -42,6 +44,10 @@ export async function scaleDown(): Promise<void> {
       },
     );
 
+    const runnersRgions = new Set<string>(
```
[the remark this answers is not in the document]
Comment (jeanschmidt) [reply terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts:47]: Thanks for pointing this out :)

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/config.ts [hunk 5]
```diff
@@ -98,6 +100,8 @@ export class Config {
...
+    this.sSMParamMaxCleanupCount = Number(process.env.SSM_PARAM_CLEANUP_AGE_DAYS || '50');
```
[the remark this answers is not in the document]
Comment (jeanschmidt) [reply terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/config.ts:104]: Thanks for pointing this out (2) :)

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/config.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeanschmidt) [reply terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/config.ts:103]: It does not, for now.

But as a great standard, I do make any relevant behaviour to be configurable by environment variables. In worst case scenario we can tune it without having to make a deployment and just going to the AWS console and editing the lambda params. And if we want to change those numbers in the future for canary vs main, we can also make terraform-only changes.

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/runners.ts [hunk 6]
```diff
@@ -352,6 +361,7 @@ async function addSSMParameterRunnerConfig(
 
 function getLaunchTemplateName(runnerParameters: RunnerInputParameters): Array<string | undefined> {
   if (runnerParameters.runnerType.os === 'linux') {
+    /* istanbul ignore next */
```
[the remark this answers is not in the document]
Comment (jeanschmidt) [reply terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/runners.ts:364]: ignore next line code or branch (if there is conditional statements in the next line for example) against code coverage tests. It is a decision in the format of: next statement is not fully tested, mostly because we don't care for a particular side effect or  anything, so don't include in the test coverage reports because we don't care.

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.test.ts [hunk 7]
```diff
@@ -1265,6 +1386,122 @@ describe('scale-down', () => {
     });
   });
 
+  describe('cleanupOldSSMParameters', () => {
+    it('Stops when LastModifiedDate === undefined', async () => {
```
[the remark this answers is not in the document]
Comment (jeanschmidt) [reply terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.test.ts:1390]: Well, this is odd, but in the AWS docs I don't find it to be possible to ever not be defined. But, in the interface description on typescript it can.

So I decided to properly treat this case, for the sake of code completeness / safeness. And decided to consider it as soo new that this value is yet to be populated.

Random guess.

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts [hunk 8]
```diff
@@ -225,6 +239,46 @@ export async function scaleDown(): Promise<void> {
   }
 }
 
+export async function cleanupOldSSMParameters(runnersRgions: Set<string>, metrics: ScaleDownMetrics): Promise<void> {
+  try {
+    for (const awsRegion of runnersRgions) {
+      const ssmParams = sortSSMParaametersByUpdateTime(
```
[the remark this answers is not in the document]
Comment (jeanschmidt) [reply terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts:245]: thanks for pointing this out (3)

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/config.ts (hunk 2 above)
Comment (jeanschmidt) [reply terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/config.ts:103]: if you read the code, [when you asked where it is used](https://github.com/pytorch/test-infra/pull/5479#issuecomment-2246149918), you'll see that it should be deleted by the instance itself when it starts.

Those are side effects of instances failing to start for a reason or another

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts (hunk 1 above)
Comment (jeanschmidt) [reply terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts:255]: I believe that this level of documentation is not necessary, justify each decision like why I decide to properly treat the correct way to handle a Typescript template even if it does not match AWS documentation is odd and too much IMO.

File: terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts (hunk 3 above)
Comment (jeanschmidt) [reply terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts:267]: We could, but I really dislike those variable names that are a phrase and very long.
Can we come up with a name that is shorter? `sSMParamMaxCleanupAllowance`? 
It is a bit too high for my taste, but I am kind of OK with that.

Having said that, can you explain what you mean by:
```
Looks like sSMParamMaxCleanupCount is the max number of params that should get cleaned up on a single scale down request. However, when ten instances are being scaled down one after another, you could end up deleting 10 * sSMParamMaxCleanupCount instances by the end.
```

Comment (ZainRizvi) [reply terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts:267]: Sure, `sSMParamMaxCleanupAllowance` works

> Having said that, can you explain what you mean by:
The parameter's functionality just surprised me.  It seems like the intention was to limit how much damage a single scale down request can do, but when there are multiple scale down requests rouge code can still do lots of damage.

Or perhaps the intention here was to limit the max duration of a scale down request? That would make more sense.

Comment (jeanschmidt) [reply terraform-aws-github-runner/modules/runners/lambdas/runners/src/scale-runners/scale-down.ts:267]: That is the intention, I did review the code and it seems it is doing as planned.

It limits the impact for `sSMParamMaxCleanupCount` by run by AWS region (we only have 1 region atm, as multi-region is a WIP)
